ferrischat_webserver/src/invites/use_invite.rs

```diff
@@ -0,0 +1,109 @@
+
+    let db = get_db_or_fail!();
+
+    let resp = sqlx::query!("SELECT * FROM invites WHERE code = $1", invite_code)
+        .fetch_optional(db)
+        .await;
+
+    match resp {
+        Ok(resp) => match resp {
+            Some(invite) => {
+                let uses = invite.uses + 1;
+                let now = {
+                    let now = time::OffsetDateTime::now_utc();
+                    time::PrimitiveDateTime::new(now.clone().date(), now.time())
+                };
+                if invite.max_uses < uses {
+                    let delete_resp =
+                        sqlx::query!("DELETE FROM invites WHERE code = $1", invite_code)
+                            .execute(db)
+                            .await;
+
+                    match delete_resp {
+                        Ok(_) => return HttpResponse::Gone().finish(),
+                        Err(e) => {
+                            return HttpResponse::InternalServerError().json(
+                                InternalServerErrorJson {
+                                    reason: format!("DB returned an error: {}", e),
+                                },
+                            )
+                        }
+                    }
+                }
+
+                if invite.max_age.is_some() {
+                    if (now - invite.created_at).whole_seconds() > invite.max_age {
+                        let delete_resp =
+                            sqlx::query!("DELETE FROM invites WHERE code = $1", invite_code)
+                                .execute(db)
+                                .await;
+
+                        match delete_resp {
+                            Ok(_) => return HttpResponse::Gone().finish(),
+                            Err(e) => {
+                                return HttpResponse::InternalServerError().json(
+                                    InternalServerErrorJson {
+                                        reason: format!("DB returned an error: {}", e),
+                                    },
+                                )
+                            }
+                        }
+                    }
+                }
+
+                let member_resp = sqlx::query!(
+                    "INSERT INTO members VALUES ($1, $2)",
+                    bigint_user_id,
+                    invite.guild_id
+                )
+                .execute(db)
+                .await;
+
+                match member_resp {
+                    Ok(_) => (),
+                    Err(e) => {
+                        return HttpResponse::InternalServerError().json(InternalServerErrorJson {
+                            reason: format!("DB returned an error: {}", e),
+                        })
+                    }
+                }
+
+                let uses_resp = sqlx::query!(
+                    "UPDATE invites SET uses = $1 WHERE code = $2",
+                    uses,
+                    invite_code
```

Comment:
My worry with this would be these queries not being atomic: if the invite were to be used twice quickly it might end up not reliably counting.
Not really any way around it, so just want to mention this.

Reply:
I have also thought about that, but yes I don't think there are anyway to work around it.